File: api/src/models/workspace.rs
use chrono::Local;
use mongodb::bson::oid::ObjectId;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, Copy)]
pub enum Visibility {
    PUBLIC,
    PRIVATE,
}


#[derive(Debug, Serialize, Deserialize)]
pub struct Workspace {
    pub _id: ObjectId,
    pub owner: ObjectId,
    pub name: String,
    pub visibility: Visibility,
    pub deleted: bool,
    pub is_active: bool,
    pub created: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WorkspaceRequest {
    pub name: String,
    pub visibility: Visibility
}

#[derive(Debug, Serialize, Deserialize)]
pub struct WorkspaceResponse {
    pub id: String,
    pub owner: String,
    pub name: String,
    pub visibility: Visibility,
    pub deleted: bool,
    pub is_active: bool,
    pub created: String,
}

impl TryFrom<WorkspaceRequest> for Workspace {
    type Error = Box<dyn std::error::Error>;

    fn try_from(item: WorkspaceRequest) -> Result<Self, Self::Error> {
        Ok(Self {
            _id: ObjectId::new(),
            owner: ObjectId::new(),
            name: item.name,
            visibility: item.visibility,
            deleted: false,
            is_active: true,
            created: Local::now().to_string(),
        })
    }
}

impl WorkspaceResponse {
    pub fn new(
        id: String,
        owner: String,
        name: String,
        visibility: Visibility,
        deleted: bool,
        is_active: bool,
        created: String,
    ) -> Self {
        WorkspaceResponse {
            id,
            owner,
            name,
            visibility,
            deleted,
            is_active,
            created,
        }
    }
}

File: app/src/types/Workspace.ts
export interface _Workspace {
  id: string;
  name: string;
  owner: string;
  is_private: boolean;
  created: Date;
  deleted: Boolean;
}

export class Workspace {
  id: string;
  name: string;
  owner: string;
  is_private: boolean;
  created: Date;
  deleted: Boolean;

  constructor(
    id: string,
    name: string,
    owner: string,
    is_private: boolean,
    created: Date,
    deleted: Boolean
  ) {
    this.id = id;
    this.name = name;
    this.owner = owner;
    this.is_private = is_private;
    this.created = created;
    this.deleted = deleted;
  }

  static createObject(obj: _Workspace) {
    return new Workspace(obj.id, obj.name, "", obj.is_private, obj.created, obj.deleted);
  }

  static createEmptyObject() {
    return new Workspace("", "", "", false, new Date(), false);
  }
}
